core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -674,135 +675,232 @@ private RexNode simplifyCoalesce(RexCall call) {
+
+    if (values.size() == 1) {
+      final RexNode firstValue = branches.get(0).value;
+      if (!call.getType().equals(firstValue.getType())) {
+        return rexBuilder.makeAbstractCast(call.getType(), firstValue);
       }
-      return last;
+      return firstValue;
     }
-  trueFalse:
+
     if (call.getType().getSqlTypeName() == SqlTypeName.BOOLEAN) {
-      // Optimize CASE where every branch returns constant true or constant
-      // false.
-      final List<Pair<RexNode, RexNode>> pairs =
-          casePairs(rexBuilder, newOperands);
-      // 1) Possible simplification if unknown is treated as false:
-      //   CASE
-      //   WHEN p1 THEN TRUE
-      //   WHEN p2 THEN TRUE
-      //   ELSE FALSE
-      //   END
-      // can be rewritten to: (p1 or p2)
-      if (unknownAs == FALSE) {
-        final List<RexNode> terms = new ArrayList<>();
-        int pos = 0;
-        for (; pos < pairs.size(); pos++) {
-          // True block
-          Pair<RexNode, RexNode> pair = pairs.get(pos);
-          if (!pair.getValue().isAlwaysTrue()) {
-            break;
-          }
-          terms.add(pair.getKey());
-        }
-        for (; pos < pairs.size(); pos++) {
-          // False block
-          Pair<RexNode, RexNode> pair = pairs.get(pos);
-          if (!pair.getValue().isAlwaysFalse()
-              && !RexUtil.isNull(pair.getValue())) {
-            break;
-          }
-        }
-        if (pos == pairs.size()) {
-          final RexNode disjunction =
-              RexUtil.composeDisjunction(rexBuilder, terms);
-          if (!call.getType().equals(disjunction.getType())) {
-            return rexBuilder.makeCast(call.getType(), disjunction);
-          }
-          return disjunction;
+      final RexNode result = simplifyBooleanCase(rexBuilder, branches, unknownAs, branchType);
+      if (result != null) {
+        if (!call.getType().equals(result.getType())) {
+          return simplify(rexBuilder.makeCast(call.getType(), result), unknownAs);
         }
+        return simplify(result, unknownAs);
       }
-      // 2) Another simplification
-      //   CASE
-      //   WHEN p1 THEN TRUE
-      //   WHEN p2 THEN FALSE
-      //   WHEN p3 THEN TRUE
-      //   ELSE FALSE
-      //   END
-      // if p1...pn cannot be nullable
-      for (Ord<Pair<RexNode, RexNode>> pair : Ord.zip(pairs)) {
-        if (pair.e.getKey().getType().isNullable()) {
-          break trueFalse;
-        }
-        if (!pair.e.getValue().isAlwaysTrue()
-            && !pair.e.getValue().isAlwaysFalse()
-            && (unknownAs == UNKNOWN || !RexUtil.isNull(pair.e.getValue()))) {
-          break trueFalse;
-        }
+    }
+    List<RexNode> newOperands = CaseBranch.toCaseOperands(rexBuilder, branches);
+    if (newOperands.equals(call.getOperands())) {
+      return call;
+    }
+    return call.clone(call.getType(), newOperands);
+  }
+
+  /** Object to describe a Case branch */
+  static class CaseBranch {
```

Comment:
If you think having a new internal class makes code more readable, maybe it can extend org.apache.calcite.util.Pair since it has some other useful methods (comparator, etc.)? Not a blocker, just a minor comment.